File: llang/L0/lib/ast_typed.ml
(* This module contains the typed abstract syntax tree for the language *)

open Symbols
open Ast

type ast_typed = 
  | Num of int
  | Bool of bool
  | BinOp of Ast.bin_op * ast_typed * ast_typed * l_type
  | UnOp of Ast.un_op * ast_typed * l_type

let type_of = function
  | Num _ -> TInt
  | Bool _ -> TBool
  | BinOp (_, _, _, t) -> t
  | UnOp (_, _, t) -> t

File: llang/L0/rust/src/ast_typed.rs
use std::rc::Rc;

use crate::symbols::Symbol;
use crate::ast::{BinOp, UnOp};
use crate::ast::{Type, TypeData};


pub type AstTyped = Rc<AstTypedData>;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum AstTypedData {
    Num(i64),
    Bool(bool),
    BinOp(BinOp, AstTyped, AstTyped, Type),
    UnOp(UnOp, AstTyped, Type),
}

impl AstTypedData {
    pub fn type_of(&self) -> Type {
        match self {
            AstTypedData::Num(_) => Rc::new(TypeData::Int),
            AstTypedData::Bool(_) => Rc::new(TypeData::Bool),
            AstTypedData::BinOp(_, _, _, t) => t.clone(),
            AstTypedData::UnOp(_, _, t) => t.clone(),
        }
    }
}